### DISCUSSION TITLE
Track if uncontrolled input is active

### DISCUSSION THREAD

**Discussion by made-in-nz** [react-hook-form/react-hook-form#8385] [Q&A] [answered]
Hi, I'm new to react-hook-form and I feel like I'm missing something obvious in my usage.
I'm writing a reusable uncontrolled input component that simply uses bootstrap 5 classes for styling that indicates success/error validation.  When the input has focus I want to remove the bootstrap 5 classes that indicate success/error.  
As an uncontrolled input, how do I essentially track an active/focused state?   I realise the API does not provide this specifically, and I have seen other discussions/issues that refer to v6 or earlier react-hook-form APIs relating to tracking onFocus/onBlur to infer an active/focused state, but I can't see how to do this with the v7 API.

**Comment by made-in-nz** [comment, marked as answer]
I think I may have the answer - simply pass an onBlur handler to register(), and set an onFocus event handler on the input?
ie:

```
const Input = ({ register, control, name, getFieldState, width, label, ...rest }) => {
  const [active, setActive] = useState(false);
  const validationState = validateState(getFieldState(name), active)  

  const handleBlur = () => {
    setActive(false);
  };

  const handleFocus = () => {
    setActive(true);
  };

  // ... some other stuff here...

  return (
    <div className="row mb-3">
      <div className={`col-md-${md}`}>
        <label htmlFor={name} className={label_class}>
          {label}
        </label>
        <input
          {...register(name, { onBlur: handleBlur })}
          className={input_class}
          onFocus={handleFocus}
          {...rest}
        />
      </div>
    </div>
  );
};
```
As I'd also like easy access to the input value, would it be better to use useController?

**Reply by made-in-nz** [reply]
Also noting I change to useController to have access to input value & error on the component.
```
// Bootstrap5 styled input with validation feedback
const RHFFormControl = ({ name: fname, control, width, label, ...rest }) => {
  const {
    field: { onChange, onBlur, name, value, ref },
    fieldState,
  } = useController({
    name: fname,
    control,
  });
  const [active, setActive] = useState(false);

  const validationState = validateState(fieldState, active, value);

  const message = validateMessage(
    fieldState,
    rest.disabled,
    rest.errorsWhileDisabled
  );

  const md = width ? width : 12;

  const input_class = cx("form-control", {
    "is-valid": validationState === "success",
    "is-invalid": validationState === "error",
  });

  const label_class = cx("form-label", {
    "text-danger": validationState === "error",
  });

  const handleBlur = (e) => {
    setActive(false);
    onBlur(e);
  };

  const handleFocus = () => {
    setActive(true);
  };

  return (
    <div className="row mb-3">
      <div className={`col-md-${md}`}>
        <label htmlFor={name} className={label_class}>
          {label}
        </label>
        <input
          onChange={onChange}
          onBlur={handleBlur}
          onFocus={handleFocus}
          name={name}
          ref={ref}
          className={input_class}
          {...rest}
        />
        {message && <div className="invalid-feedback">{message}</div>}
      </div>
    </div>
  );
};
```

**Comment by Moshyfawn** [comment]
Maybe you're looking for [`mode: onTouched`](https://react-hook-form.com/api/useform/#main=) instead?

**Reply by made-in-nz** [reply]
I'm not sure how `mode: onTouched` would help - that's a validation strategy right? 
I have server side (async) validation, so have set up my form with a `resolver` function that handles the async validation, and `mode: onBlur`.  What I'm trying to achieve is this:
1. User enters a field & enters invalid data.  After blur & async validation UI updates field to indicate error/success (eg red/green colour).
2. User re-enters invalid field.  UI clears error/success state while field has focus.
3. After blur but during async validation, UI does not update field state.  Once async validation completes, UI updates field with error/success state.

To achieve this (and as a separate issue) I'd also like to track an 'isValidating' state per field so that the UI does not update the validation state of the input until the async validation has completed, otherwise UI flickers between states during async validation. 

I've seen issues raised about tracking an isActive & isValidating state per field here, but neither progressed.  I'm assuming others, like me, are migrating from redux-form where this was provided by the API.

My solution above works for tracking isActive, so I'll mark it as the solution.  Thanks!
